pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
 
 	return kadm.GroupMemberLag{}, lastErr
 }
+
+type partitionState struct {
+	Partition          int32
+	Lag                int64
+	Commit             kadm.Offset
+	CommitRecTimestamp time.Time
+	SeenTillOffset     int64
+	LastBlockEnd       int64
+}
+
+func partitionStateFromLag(logger log.Logger, lag kadm.GroupMemberLag, fallbackMillis int64) *partitionState {
+	commitRecTs, seenTillOffset, lastBlockEnd, err := unmarshallCommitMeta(lag.Commit.Metadata)
+	if err != nil {
+		// If there is an error in unmarshalling the metadata, treat it as if
+		// we have no commit. There is no reason to stop the cycle for this.
+		level.Warn(logger).Log("msg", "error unmarshalling commit metadata", "err", err, "partition", lag.Partition, "offset", lag.Commit.At, "metadata", lag.Commit.Metadata)
+	}
+
+	if commitRecTs == 0 {
+		// If there was no commit metadata, we use the fallback to replay a set amount of
+		// records because it is non-trivial to peek at the first record in a partition to determine
+		// the range of replay required. Without knowing the range, we might end up trying to consume
+		// a lot of records in a single partition consumption call and end up in an OOM loop.
+		// TODO(codesome): add a test for this.
+		commitRecTs = fallbackMillis
+	}
+
+	return &partitionState{
+		Partition:          lag.Partition,
+		Lag:                lag.Lag,
+		Commit:             lag.Commit,
+		CommitRecTimestamp: time.UnixMilli(commitRecTs),
+		SeenTillOffset:     seenTillOffset,
+		LastBlockEnd:       lastBlockEnd,
+	}
+}
+
+func (b *BlockBuilder) consumePartition(ctx context.Context, state *partitionState, cycleEnd time.Time) error {
+	lagging := cycleEnd.Sub(state.CommitRecTimestamp) > 3*b.cfg.ConsumeInterval/2
+	if !lagging {
+		// Either we did not find a commit offset or we are not lagging behind by
+		// more than 1.5 times the consume interval.
+		// When there is no kafka commit, we play safe and assume seenTillOffset and
+		// lastBlockEnd was 0 to not discard any samples unnecessarily.
+		err := b.consumePartitionCycle(ctx, state, cycleEnd)
```

Comment:
if there's no commit don't we want to still chop up the period into multiple cycles? I mean do we want to compact the last 10 hours in 1 sections or in 10 sections?

Which also leads to another question: what's the difference between consuming only the last section and consuming multiple sections? i.e. why do we need this `if`, can't we just use the loop below?

Reply:
The problem is that the `cycleEnd` is relative to the record's timestamp — not the timestamp of the commit (i.e. when distributor pushed the original record to kafka). If we haven't committed the metadata, that contains `CommitRecTimestamp`, we cannot tell the timestamp of the first record, w/o consuming it. Does it make sense?

Reply:
I recall that, initially, we tried chopping the cycle from inside the consumption loop:

```
for recs := range kafka.Consume() {
  for rec := range recs {
    if rec.Ts > cycleEnd {
      return
    }
    if cycleEnd - rec.Ts > interval {
      tsdb.compact()
      tsdb.upload()
      kafka.commit()

      cycleEnd = rec.Ts + interval
    }
    ...
  }
}
```

But that kept showing edge cases, and we've refactored it the current form (with keeping the `rec.Ts` in the commit's metadata to make it simpler). I'm not a fan of this one, but it works. Maybe we can improve it now 🙃

Reply:
i like how you did it now. I had basically something like that in mind.